File: sqlite/src/memory_vfs.c
#include "memory_vfs.h"
#include <stdlib.h>
#include <string.h>

/* Intentionally portable C: the same adapter is compiled with the native oracle
 * and translated with SQLite. Allocation/memory calls use dotcc's libc port. */
#define MEM_PATH_MAX 1024
#define MEM_I64_MAX 9223372036854775807LL

typedef struct MemNode MemNode;
typedef struct MemFile MemFile;
struct MemNode {
    char *name;
    unsigned char *data;
    sqlite3_int64 size;
    size_t capacity;
    int refs;
    int unlinked;
    MemNode *next;
};
struct MemFile {
    sqlite3_file base;
    MemNode *node;
    int flags;
    int lock;
    MemFile *next;
};

static MemNode *mem_nodes;
static MemFile *mem_handles;
static int mem_fail_mask;
static int mem_fail_count;
static int mem_fail_result;
static unsigned int mem_random = 1;
static sqlite3_int64 mem_unix_ms = 1735689600000LL; /* 2025-01-01 UTC */

static int mem_failure(int operation) {
    if ((mem_fail_mask & operation) == 0) return SQLITE_OK;
    if (mem_fail_count > 0) {
        --mem_fail_count;
        return SQLITE_OK;
    }
    mem_fail_mask = 0;
    return mem_fail_result;
}

static int mem_path(const char *name, int count, char *out) {
    int used = 1;
    const char *start;
    size_t len;
    if (!name || count < 2) return SQLITE_CANTOPEN;
    out[0] = '/';
    while (*name) {
        while (*name == '/') ++name;
        start = name;
        while (*name && *name != '/') ++name;
        len = (size_t)(name - start);
        if (len == 0 || (len == 1 && start[0] == '.')) continue;
        if (len == 2 && start[0] == '.' && start[1] == '.') {
            while (used > 1 && out[used - 1] != '/') --used;
            if (used > 1) --used;
            continue;
        }
        if (len > MEM_PATH_MAX || used + (int)len + (used > 1) >= count)
            return SQLITE_CANTOPEN;
        if (used > 1) out[used++] = '/';
        memcpy(out + used, start, len);
        used += (int)len;
    }
    out[used] = 0;
    return SQLITE_OK;
}

static MemNode *mem_find(const char *name) {
    MemNode *node;
    for (node = mem_nodes; node; node = node->next)
        if (!node->unlinked && node->name && strcmp(name, node->name) == 0)
            return node;
    return NULL;
}

static MemNode *mem_create(const char *name) {
    MemNode *node = (MemNode *)malloc(sizeof(MemNode));
    if (!node) return NULL;
    memset(node, 0, sizeof(MemNode));
    if (name) {
        node->name = (char *)malloc(strlen(name) + 1);
        if (!node->name) { free(node); return NULL; }
        strcpy(node->name, name);
    } else {
        node->unlinked = 1;
    }
    node->next = mem_nodes;
    mem_nodes = node;
    return node;
}

static void mem_destroy(MemNode *node) {
    MemNode **link = &mem_nodes;
    while (*link && *link != node) link = &(*link)->next;
    if (*link) *link = node->next;
    free(node->name);
    free(node->data);
    free(node);
}

static int mem_resize(MemNode *node, sqlite3_int64 size) {
    unsigned char *data;
    size_t capacity;
    if (size < 0 || (sqlite3_uint64)size > (size_t)-1) return SQLITE_FULL;
    if ((size_t)size > node->capacity) {
        capacity = (size_t)size;
        if (capacity <= ((size_t)-1) - 4095)
            capacity = ((capacity + 4095) / 4096) * 4096;
        data = (unsigned char *)realloc(node->data, capacity);
        if (!data) return SQLITE_NOMEM;
        node->data = data;
        node->capacity = capacity;
    }
    if (size > node->size)
        memset(node->data + (size_t)node->size, 0, (size_t)(size - node->size));
    node->size = size;
    return SQLITE_OK;
}

static int mem_close(sqlite3_file *base) {
    MemFile *file = (MemFile *)base;
    MemFile **link = &mem_handles;
    MemNode *node = file->node;
    while (*link && *link != file) link = &(*link)->next;
    if (*link) *link = file->next;
    if (file->flags & SQLITE_OPEN_DELETEONCLOSE) node->unlinked = 1;
    --node->refs;
    if (!node->refs && node->unlinked) mem_destroy(node);
    base->pMethods = NULL;
    return SQLITE_OK;
}

static int mem_read(sqlite3_file *base, void *buffer, int count, sqlite3_int64 offset) {
    MemNode *node = ((MemFile *)base)->node;
    sqlite3_int64 available;
    int rc = mem_failure(DOTCC_VFS_FAIL_READ);
    if (rc != SQLITE_OK) return rc;
    if (offset < 0 || count < 0) return SQLITE_IOERR_READ;
    if (!count) return SQLITE_OK;
    available = offset < node->size ? node->size - offset : 0;
    if (available > count) available = count;
    if (available) memcpy(buffer, node->data + (size_t)offset, (size_t)available);
    if (available < count) {
        memset((unsigned char *)buffer + (size_t)available, 0, (size_t)(count - available));
        return SQLITE_IOERR_SHORT_READ;
    }
    return SQLITE_OK;
}

static int mem_write(sqlite3_file *base, const void *buffer, int count, sqlite3_int64 offset) {
    MemFile *file = (MemFile *)base;
    int rc;
    if (file->flags & SQLITE_OPEN_READONLY) return SQLITE_READONLY;
    rc = mem_failure(DOTCC_VFS_FAIL_WRITE);
    if (rc != SQLITE_OK) return rc;
    if (offset < 0 || count < 0 || offset > MEM_I64_MAX - count) return SQLITE_IOERR_WRITE;
    if (!count) return SQLITE_OK;
    if (offset + count > file->node->size) {
        rc = mem_resize(file->node, offset + count);
        if (rc != SQLITE_OK) return rc;
    }
    memcpy(file->node->data + (size_t)offset, buffer, (size_t)count);
    return SQLITE_OK;
}

static int mem_truncate(sqlite3_file *base, sqlite3_int64 size) {
    MemFile *file = (MemFile *)base;
    int rc;
    if (file->flags & SQLITE_OPEN_READONLY) return SQLITE_READONLY;
    rc = mem_failure(DOTCC_VFS_FAIL_TRUNCATE);
    if (rc != SQLITE_OK) return rc;
    if (size < 0) return SQLITE_IOERR_TRUNCATE;
    return mem_resize(file->node, size);
}

static int mem_sync(sqlite3_file *base, int flags) {
    (void)base; (void)flags;
    return mem_failure(DOTCC_VFS_FAIL_SYNC);
}

static int mem_size(sqlite3_file *base, sqlite3_int64 *size) {
    *size = ((MemFile *)base)->node->size;
    return SQLITE_OK;
}

static int mem_lock(sqlite3_file *base, int level) {
    MemFile *file = (MemFile *)base;
    MemFile *other;
    if (level <= file->lock) return SQLITE_OK;
    if (level < SQLITE_LOCK_SHARED || level > SQLITE_LOCK_EXCLUSIVE) return SQLITE_IOERR_LOCK;
    for (other = mem_handles; other; other = other->next) {
        if (other == file || other->node != file->node) continue;
        if ((level == SQLITE_LOCK_SHARED && other->lock >= SQLITE_LOCK_PENDING) ||
            (level >= SQLITE_LOCK_RESERVED && other->lock >= SQLITE_LOCK_RESERVED))
            return SQLITE_BUSY;
    }
    if (level == SQLITE_LOCK_EXCLUSIVE) {
        /* Retain PENDING on contention to prevent starvation from new readers. */
        file->lock = SQLITE_LOCK_PENDING;
        for (other = mem_handles; other; other = other->next)
            if (other != file && other->node == file->node && other->lock >= SQLITE_LOCK_SHARED)
                return SQLITE_BUSY;
    }
    file->lock = level;
    return SQLITE_OK;
}

static int mem_unlock(sqlite3_file *base, int level) {
    MemFile *file = (MemFile *)base;
    if (level != SQLITE_LOCK_NONE && level != SQLITE_LOCK_SHARED) return SQLITE_IOERR_UNLOCK;
    if (level < file->lock) file->lock = level;
    return SQLITE_OK;
}

static int mem_reserved(sqlite3_file *base, int *result) {
    MemFile *file = (MemFile *)base;
    MemFile *other;
    *result = 0;
    for (other = mem_handles; other; other = other->next)
        if (other->node == file->node && other->lock >= SQLITE_LOCK_RESERVED) *result = 1;
    return SQLITE_OK;
}

static int mem_control(sqlite3_file *base, int operation, void *argument) {
    if (operation == SQLITE_FCNTL_LOCKSTATE) {
        *(int *)argument = ((MemFile *)base)->lock;
        return SQLITE_OK;
    }
    return SQLITE_NOTFOUND;
}

static int mem_sector(sqlite3_file *base) { (void)base; return 512; }
static int mem_device(sqlite3_file *base) { (void)base; return 0; }

static const sqlite3_io_methods mem_methods = {
    1, mem_close, mem_read, mem_write, mem_truncate, mem_sync, mem_size,
    mem_lock, mem_unlock, mem_reserved, mem_control, mem_sector, mem_device,
    NULL, NULL, NULL, NULL, NULL, NULL
};

static int mem_open(sqlite3_vfs *vfs, const char *name, sqlite3_file *base,
                    int flags, int *out_flags) {
    MemFile *file = (MemFile *)base;
    MemNode *node;
    char path[MEM_PATH_MAX + 1];
    int rc;
    (void)vfs;
    memset(file, 0, sizeof(MemFile));
    rc = mem_failure(DOTCC_VFS_FAIL_OPEN);
    if (rc != SQLITE_OK) return rc;
    if (!(flags & (SQLITE_OPEN_READONLY | SQLITE_OPEN_READWRITE)) ||
        ((flags & SQLITE_OPEN_READONLY) && (flags & SQLITE_OPEN_READWRITE)) ||
        ((flags & SQLITE_OPEN_CREATE) && !(flags & SQLITE_OPEN_READWRITE)))
        return SQLITE_CANTOPEN;
    if (name) {
        rc = mem_path(name, sizeof(path), path);
        if (rc != SQLITE_OK) return rc;
        node = mem_find(path);
        if (node && (flags & SQLITE_OPEN_EXCLUSIVE) && (flags & SQLITE_OPEN_CREATE))
            return SQLITE_CANTOPEN;
        if (!node && !(flags & SQLITE_OPEN_CREATE)) return SQLITE_CANTOPEN;
        if (!node) node = mem_create(path);
    } else {
        if (!(flags & SQLITE_OPEN_DELETEONCLOSE)) return SQLITE_CANTOPEN;
        node = mem_create(NULL);
    }
    if (!node) return SQLITE_NOMEM;
    file->node = node;
    file->flags = flags;
    file->next = mem_handles;
    mem_handles = file;
    ++node->refs;
    base->pMethods = &mem_methods;
    if (out_flags) *out_flags = flags;
    return SQLITE_OK;
}

static int mem_delete(sqlite3_vfs *vfs, const char *name, int sync_dir) {
    char path[MEM_PATH_MAX + 1];
    MemNode *node;
    int rc = mem_failure(DOTCC_VFS_FAIL_DELETE);
    (void)vfs; (void)sync_dir;
    if (rc != SQLITE_OK) return rc;
    rc = mem_path(name, sizeof(path), path);
    if (rc != SQLITE_OK) return rc;
    node = mem_find(path);
    if (!node) return SQLITE_IOERR_DELETE_NOENT;
    node->unlinked = 1;
    if (!node->refs) mem_destroy(node);
    return SQLITE_OK;
}

static int mem_access(sqlite3_vfs *vfs, const char *name, int flags, int *result) {
    char path[MEM_PATH_MAX + 1];
    int rc;
    (void)vfs;
    *result = 0;
    if (flags != SQLITE_ACCESS_EXISTS && flags != SQLITE_ACCESS_READ && flags != SQLITE_ACCESS_READWRITE)
        return SQLITE_IOERR_ACCESS;
    rc = mem_path(name, sizeof(path), path);
    if (rc != SQLITE_OK) return rc;
    *result = mem_find(path) != NULL;
    return SQLITE_OK;
}

static int mem_fullpath(sqlite3_vfs *vfs, const char *name, int count, char *out) {
    (void)vfs;
    return mem_path(name, count, out);
}

static int mem_randomness(sqlite3_vfs *vfs, int count, char *out) {
    int i;
    (void)vfs;
    for (i = 0; i < count; ++i) {
        mem_random = mem_random * 1664525U + 1013904223U;
        out[i] = (char)(mem_random >> 24);
    }
    return count;
}

static int mem_sleep(sqlite3_vfs *vfs, int microseconds) {
    (void)vfs;
    if (microseconds > 0) mem_unix_ms += microseconds / 1000;
    return microseconds > 0 ? microseconds : 0;
}

static int mem_time(sqlite3_vfs *vfs, double *julian_day) {
    (void)vfs;
    *julian_day = 2440587.5 + (double)mem_unix_ms / 86400000.0;
    return SQLITE_OK;
}

static int mem_error(sqlite3_vfs *vfs, int count, char *out) {
    (void)vfs;
    if (count > 0) out[0] = 0;
    return 0;
}

static sqlite3_vfs mem_vfs = {
    1, sizeof(MemFile), MEM_PATH_MAX, NULL, DOTCC_MEMORY_VFS_NAME, NULL,
    mem_open, mem_delete, mem_access, mem_fullpath,
    NULL, NULL, NULL, NULL, /* Dynamic loading is deliberately unsupported. */
    mem_randomness, mem_sleep, mem_time, mem_error,
    NULL, NULL, NULL, NULL
};

sqlite3_vfs *dotcc_memory_vfs(void) { return &mem_vfs; }
int sqlite3_os_init(void) { return sqlite3_vfs_register(&mem_vfs, 1); }
int sqlite3_os_end(void) { return sqlite3_vfs_unregister(&mem_vfs); }

int dotcc_memory_vfs_reset(void) {
    if (mem_handles) return SQLITE_BUSY;
    while (mem_nodes) mem_destroy(mem_nodes);
    mem_fail_mask = 0;
    mem_random = 1;
    mem_unix_ms = 1735689600000LL;
    return SQLITE_OK;
}

int dotcc_memory_vfs_file_count(void) {
    MemNode *node;
    int count = 0;
    for (node = mem_nodes; node; node = node->next) ++count;
    return count;
}

int dotcc_memory_vfs_handle_count(void) {
    MemFile *file;
    int count = 0;
    for (file = mem_handles; file; file = file->next) ++count;
    return count;
}

void dotcc_memory_vfs_fail_after(int mask, int successful_calls, int result) {
    mem_fail_mask = mask;
    mem_fail_count = successful_calls > 0 ? successful_calls : 0;
    mem_fail_result = result == SQLITE_OK ? SQLITE_IOERR : result;
}

void dotcc_memory_vfs_seed(unsigned int seed) { mem_random = seed; }
void dotcc_memory_vfs_time(sqlite3_int64 unix_milliseconds) { mem_unix_ms = unix_milliseconds; }

int dotcc_memory_vfs_export(const char *name, void *buffer,
                            sqlite3_int64 capacity, sqlite3_int64 *size) {
    char path[MEM_PATH_MAX + 1];
    MemNode *node;
    int rc;
    if (!size || capacity < 0) return SQLITE_MISUSE;
    rc = mem_path(name, sizeof(path), path);
    if (rc != SQLITE_OK) return rc;
    node = mem_find(path);
    if (!node) return SQLITE_NOTFOUND;
    if (node->refs) return SQLITE_BUSY;
    *size = node->size;
    if (!buffer) return SQLITE_OK;
    if (capacity < node->size) return SQLITE_TOOBIG;
    if (node->size) memcpy(buffer, node->data, (size_t)node->size);
    return SQLITE_OK;
}

int dotcc_memory_vfs_import(const char *name, const void *buffer, sqlite3_int64 size) {
    char path[MEM_PATH_MAX + 1];
    MemNode *node;
    int created = 0;
    int rc;
    if (size < 0 || (!buffer && size)) return SQLITE_MISUSE;
    rc = mem_path(name, sizeof(path), path);
    if (rc != SQLITE_OK) return rc;
    node = mem_find(path);
    if (node && node->refs) return SQLITE_BUSY;
    if (!node) { node = mem_create(path); created = 1; }
    if (!node) return SQLITE_NOMEM;
    rc = mem_resize(node, size);
    if (rc != SQLITE_OK) {
        if (created) mem_destroy(node);
        return rc;
    }
    if (size) memcpy(node->data, buffer, (size_t)size);
    return SQLITE_OK;
}

File: sqlite/src/MemoryVfs.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
#if DOTCC_SQLITE_PRODUCT
using static Managed.Database.Sqlite;
#else
using static DotCcProgram;
#endif

namespace Managed.Database;

/// <summary>Process-local, deterministic SQLite VFS. No disk, WAL or mmap support.</summary>
internal static unsafe class MemoryVfs
{
    private const int PathBytes = 1025;
    private const int FailOpen = 1, FailRead = 2, FailWrite = 4, FailTruncate = 8, FailSync = 16, FailDelete = 32;
    private const long InitialTime = 1735689600000;
    // No SQLite allocator or mutex call while holding this gate. Initialization,
    // failure injection and cold restart therefore cannot invert SQLite's locks.
    private static readonly object Gate = new();
    private static readonly List<Node> Nodes = [];
    private static readonly Dictionary<nint, OpenFile> Handles = [];
    private static int _failMask, _failCount, _failResult;
    private static uint _random = 1;
    private static long _unixMilliseconds = InitialTime;
    private static sqlite3_vfs* _vfs;
    private static sqlite3_io_methods* _methods;

    private sealed class Node(string? name)
    {
        public readonly string? Name = name;
        public byte[] Data = [];
        public int Size, References;
        public bool Unlinked = name == null;
    }
    private sealed class OpenFile(Node node, int flags)
    {
        public readonly Node Node = node;
        public readonly int Flags = flags;
        public int LockLevel;
    }

    // Capture function addresses once; preserve their exact bits across GC/tiering.
    private static readonly sqlite3_io_methods Methods = new()
    {
        iVersion = 1, xClose = &Close, xRead = &Read, xWrite = &Write,
        xTruncate = &Truncate, xSync = &Sync, xFileSize = &Size,
        xLock = &LockFile, xUnlock = &Unlock, xCheckReservedLock = &Reserved,
        xFileControl = &Control, xSectorSize = &Sector, xDeviceCharacteristics = &Device,
    };
    private static readonly sqlite3_vfs Vfs = new()
    {
        iVersion = 1, szOsFile = sizeof(sqlite3_file), mxPathname = PathBytes - 1,
        xOpen = &Open, xDelete = &Delete, xAccess = &Access, xFullPathname = &FullPath,
        xRandomness = &Randomness, xSleep = &Sleep, xCurrentTime = &Time, xGetLastError = &Error,
    };

    public static sqlite3_vfs* GetVfs()
    {
        lock (Gate)
        {
            if (_vfs != null) return _vfs;
            sqlite3_io_methods* methods = null;
            sqlite3_vfs* vfs = null;
            byte* name = null;
            try
            {
                methods = (sqlite3_io_methods*)NativeMemory.Alloc((nuint)sizeof(sqlite3_io_methods));
                vfs = (sqlite3_vfs*)NativeMemory.Alloc((nuint)sizeof(sqlite3_vfs));
                name = (byte*)NativeMemory.Alloc(13);
                *methods = Methods;
                *vfs = Vfs;
                "dotcc-memory\0"u8.CopyTo(new Span<byte>(name, 13));
                vfs->zName = name;
                _methods = methods;
                _vfs = vfs;
                return vfs;
            }
            catch (OutOfMemoryException) { return null; }
            finally
            {
                if (_vfs == null)
                {
                    NativeMemory.Free(methods);
                    NativeMemory.Free(vfs);
                    NativeMemory.Free(name);
                }
            }
        }
    }

    public static int Initialize()
    {
        var vfs = GetVfs();
        if (vfs == null) return SQLITE_NOMEM;
#if DOTCC_SQLITE_PRODUCT
        int rc = sqlite3_vfs_register(vfs, 0);
        if (rc != SQLITE_OK) return rc;
        rc = HostVfs.RegisterVfs();
        if (rc != SQLITE_OK) sqlite3_vfs_unregister(vfs);
        return rc;
#else
        return sqlite3_vfs_register(vfs, 1);
#endif
    }

    public static int Shutdown()
    {
#if DOTCC_SQLITE_PRODUCT
        int rc = HostVfs.UnregisterVfs();
        if (rc != SQLITE_OK) return rc;
#endif
        // SQLite registrations may retain the canonical table addresses. Keep
        // the small tables for the process lifetime, as HostVfs does on restart.
        return _vfs == null ? SQLITE_OK : sqlite3_vfs_unregister(_vfs);
    }

    private static int Failure(int operation)
    {
        if ((_failMask & operation) == 0) return SQLITE_OK;
        if (_failCount > 0) { --_failCount; return SQLITE_OK; }
        _failMask = 0;
        return _failResult;
    }

    private static int Normalize(byte* name, Span<byte> output, out int length)
    {
        length = 0;
        if (name == null || output.Length < 2) return SQLITE_CANTOPEN;
        int used = 1;
        output[0] = (byte)'/';
        while (*name != 0)
        {
            while (*name == '/') ++name;
            byte* start = name;
            while (*name != 0 && *name != '/') ++name;
            long size = name - start;
            if (size == 0 || (size == 1 && start[0] == '.')) continue;
            if (size == 2 && start[0] == '.' && start[1] == '.')
            {
                while (used > 1 && output[used - 1] != '/') --used;
                if (used > 1) --used;
                continue;
            }
            if (size > PathBytes - 1 || used + size + (used > 1 ? 1 : 0) >= output.Length)
                return SQLITE_CANTOPEN;
            if (used > 1) output[used++] = (byte)'/';
            new ReadOnlySpan<byte>(start, (int)size).CopyTo(output[used..]);
            used += (int)size;
        }
        output[used] = 0;
        length = used;
        return SQLITE_OK;
    }

    private static int PathKey(byte* name, out string key)
    {
        Span<byte> path = stackalloc byte[PathBytes];
        int rc = Normalize(name, path, out int length);
        // Preserve arbitrary pathname bytes, including invalid UTF-8.
        key = rc == SQLITE_OK ? Convert.ToHexString(path[..length]) : "";
        return rc;
    }
    private static Node? Find(string name)
    {
        foreach (var node in Nodes)
            if (!node.Unlinked && node.Name == name) return node;
        return null;
    }
    private static int Resize(Node node, long size)
    {
        if (size < 0 || size > Array.MaxLength) return SQLITE_FULL;
        int length = (int)size;
        if (length > node.Data.Length)
        {
            int capacity = (int)Math.Min(((size + 4095) / 4096) * 4096, Array.MaxLength);
            Array.Resize(ref node.Data, capacity);
        }
        if (length > node.Size) Array.Clear(node.Data, node.Size, length - node.Size);
        node.Size = length;
        return SQLITE_OK;
    }

    private static int Open(sqlite3_vfs* vfs, byte* name, sqlite3_file* file, int flags, int* outputFlags)
    {
        file->pMethods = null;
        lock (Gate)
        {
            Node? created = null;
            try
            {
                int rc = Failure(FailOpen);
                if (rc != SQLITE_OK) return rc;
                if ((flags & (SQLITE_OPEN_READONLY | SQLITE_OPEN_READWRITE)) == 0 ||
                    (flags & (SQLITE_OPEN_READONLY | SQLITE_OPEN_READWRITE)) == (SQLITE_OPEN_READONLY | SQLITE_OPEN_READWRITE) ||
                    ((flags & SQLITE_OPEN_CREATE) != 0 && (flags & SQLITE_OPEN_READWRITE) == 0))
                    return SQLITE_CANTOPEN;
                Node? node;
                if (name != null)
                {
                    rc = PathKey(name, out string path);
                    if (rc != SQLITE_OK) return rc;
                    node = Find(path);
                    if (node != null && (flags & (SQLITE_OPEN_EXCLUSIVE | SQLITE_OPEN_CREATE)) == (SQLITE_OPEN_EXCLUSIVE | SQLITE_OPEN_CREATE))
                        return SQLITE_CANTOPEN;
                    if (node == null && (flags & SQLITE_OPEN_CREATE) == 0) return SQLITE_CANTOPEN;
                    if (node == null) node = created = new Node(path);
                }
                else
                {
                    if ((flags & SQLITE_OPEN_DELETEONCLOSE) == 0) return SQLITE_CANTOPEN;
                    node = created = new Node(null);
                }
                if (created != null) Nodes.Add(created);
                Handles.Add((nint)file, new OpenFile(node, flags));
                ++node.References;
                file->pMethods = _methods;
                if (outputFlags != null) *outputFlags = flags;
                return SQLITE_OK;
            }
            catch (OutOfMemoryException)
            {
                if (created != null) Nodes.Remove(created);
                return SQLITE_NOMEM;
            }
        }
    }

    private static int Close(sqlite3_file* file)
    {
        lock (Gate)
        {
            var state = Handles[(nint)file];
            Handles.Remove((nint)file);
            if ((state.Flags & SQLITE_OPEN_DELETEONCLOSE) != 0) state.Node.Unlinked = true;
            if (--state.Node.References == 0 && state.Node.Unlinked) Nodes.Remove(state.Node);
            file->pMethods = null;
            return SQLITE_OK;
        }
    }
    private static int Read(sqlite3_file* file, void* buffer, int count, long offset)
    {
        lock (Gate)
        {
            int rc = Failure(FailRead);
            if (rc != SQLITE_OK) return rc;
            if (offset < 0 || count < 0) return SQLITE_IOERR_READ;
            var node = Handles[(nint)file].Node;
            int available = (int)Math.Min(count, offset < node.Size ? node.Size - offset : 0);
            var target = new Span<byte>(buffer, count);
            if (available != 0) node.Data.AsSpan((int)offset, available).CopyTo(target);
            target[available..].Clear();
            return available < count ? SQLITE_IOERR_SHORT_READ : SQLITE_OK;
        }
    }
    private static int Write(sqlite3_file* file, void* buffer, int count, long offset)
    {
        lock (Gate)
        {
            try
            {
                var state = Handles[(nint)file];
                if ((state.Flags & SQLITE_OPEN_READONLY) != 0) return SQLITE_READONLY;
                int rc = Failure(FailWrite);
                if (rc != SQLITE_OK) return rc;
                if (offset < 0 || count < 0 || offset > long.MaxValue - count) return SQLITE_IOERR_WRITE;
                if (count == 0) return SQLITE_OK;
                if (offset + count > state.Node.Size)
                {
                    rc = Resize(state.Node, offset + count);
                    if (rc != SQLITE_OK) return rc;
                }
                new ReadOnlySpan<byte>(buffer, count).CopyTo(state.Node.Data.AsSpan((int)offset, count));
                return SQLITE_OK;
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
    private static int Truncate(sqlite3_file* file, long size)
    {
        lock (Gate)
        {
            try
            {
                var state = Handles[(nint)file];
                if ((state.Flags & SQLITE_OPEN_READONLY) != 0) return SQLITE_READONLY;
                int rc = Failure(FailTruncate);
                return rc != SQLITE_OK ? rc : size < 0 ? SQLITE_IOERR_TRUNCATE : Resize(state.Node, size);
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
    private static int Sync(sqlite3_file* file, int flags) { lock (Gate) return Failure(FailSync); }
    private static int Size(sqlite3_file* file, long* size)
    {
        lock (Gate) { *size = Handles[(nint)file].Node.Size; return SQLITE_OK; }
    }
    private static int LockFile(sqlite3_file* file, int level)
    {
        lock (Gate)
        {
            var state = Handles[(nint)file];
            if (level <= state.LockLevel) return SQLITE_OK;
            if (level < SQLITE_LOCK_SHARED || level > SQLITE_LOCK_EXCLUSIVE) return SQLITE_IOERR_LOCK;
            foreach (var other in Handles.Values)
            {
                if (other == state || other.Node != state.Node) continue;
                if ((level == SQLITE_LOCK_SHARED && other.LockLevel >= SQLITE_LOCK_PENDING) ||
                    (level >= SQLITE_LOCK_RESERVED && other.LockLevel >= SQLITE_LOCK_RESERVED)) return SQLITE_BUSY;
            }
            if (level == SQLITE_LOCK_EXCLUSIVE)
            {
                state.LockLevel = SQLITE_LOCK_PENDING;
                foreach (var other in Handles.Values)
                    if (other != state && other.Node == state.Node && other.LockLevel >= SQLITE_LOCK_SHARED)
                        return SQLITE_BUSY;
            }
            state.LockLevel = level;
            return SQLITE_OK;
        }
    }
    private static int Unlock(sqlite3_file* file, int level)
    {
        lock (Gate)
        {
            if (level != SQLITE_LOCK_NONE && level != SQLITE_LOCK_SHARED) return SQLITE_IOERR_UNLOCK;
            var state = Handles[(nint)file];
            if (level < state.LockLevel) state.LockLevel = level;
            return SQLITE_OK;
        }
    }
    private static int Reserved(sqlite3_file* file, int* result)
    {
        lock (Gate)
        {
            var node = Handles[(nint)file].Node;
            *result = 0;
            foreach (var other in Handles.Values)
                if (other.Node == node && other.LockLevel >= SQLITE_LOCK_RESERVED) *result = 1;
            return SQLITE_OK;
        }
    }
    private static int Control(sqlite3_file* file, int operation, void* argument)
    {
        lock (Gate)
        {
            if (operation != SQLITE_FCNTL_LOCKSTATE) return SQLITE_NOTFOUND;
            *(int*)argument = Handles[(nint)file].LockLevel;
            return SQLITE_OK;
        }
    }
    private static int Sector(sqlite3_file* file) => 512;
    private static int Device(sqlite3_file* file) => 0;

    private static int Delete(sqlite3_vfs* vfs, byte* name, int syncDirectory)
    {
        lock (Gate)
        {
            try
            {
                int rc = Failure(FailDelete);
                if (rc != SQLITE_OK) return rc;
                rc = PathKey(name, out string path);
                if (rc != SQLITE_OK) return rc;
                var node = Find(path);
                if (node == null) return SQLITE_IOERR_DELETE_NOENT;
                node.Unlinked = true;
                if (node.References == 0) Nodes.Remove(node);
                return SQLITE_OK;
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
    private static int Access(sqlite3_vfs* vfs, byte* name, int flags, int* result)
    {
        lock (Gate)
        {
            *result = 0;
            try
            {
                if (flags != SQLITE_ACCESS_EXISTS && flags != SQLITE_ACCESS_READ && flags != SQLITE_ACCESS_READWRITE)
                    return SQLITE_IOERR_ACCESS;
                int rc = PathKey(name, out string path);
                if (rc == SQLITE_OK) *result = Find(path) == null ? 0 : 1;
                return rc;
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
    private static int FullPath(sqlite3_vfs* vfs, byte* name, int count, byte* output) =>
        count < 2 ? SQLITE_CANTOPEN : Normalize(name, new Span<byte>(output, count), out _);
    private static int Randomness(sqlite3_vfs* vfs, int count, byte* output)
    {
        lock (Gate)
        {
            for (int i = 0; i < count; i++)
            {
                _random = unchecked(_random * 1664525U + 1013904223U);
                output[i] = (byte)(_random >> 24);
            }
            return count;
        }
    }
    private static int Sleep(sqlite3_vfs* vfs, int microseconds)
    {
        lock (Gate)
        {
            if (microseconds > 0) _unixMilliseconds = unchecked(_unixMilliseconds + microseconds / 1000);
            return Math.Max(0, microseconds);
        }
    }
    private static int Time(sqlite3_vfs* vfs, double* julianDay)
    {
        lock (Gate) { *julianDay = 2440587.5 + _unixMilliseconds / 86400000.0; return SQLITE_OK; }
    }
    private static int Error(sqlite3_vfs* vfs, int count, byte* output)
    {
        if (count > 0) output[0] = 0;
        return 0;
    }

    public static int Reset()
    {
        lock (Gate)
        {
            if (Handles.Count != 0) return SQLITE_BUSY;
            Nodes.Clear();
            _failMask = 0;
            _random = 1;
            _unixMilliseconds = InitialTime;
            return SQLITE_OK;
        }
    }
    public static int FileCount() { lock (Gate) return Nodes.Count; }
    public static int HandleCount() { lock (Gate) return Handles.Count; }
    public static void FailAfter(int mask, int successfulCalls, int result)
    {
        lock (Gate)
        {
            _failMask = mask;
            _failCount = Math.Max(0, successfulCalls);
            _failResult = result == SQLITE_OK ? SQLITE_IOERR : result;
        }
    }
    public static void Seed(uint seed) { lock (Gate) _random = seed; }
    public static void SetTime(long unixMilliseconds) { lock (Gate) _unixMilliseconds = unixMilliseconds; }
    public static int Export(byte* name, void* buffer, long capacity, long* size)
    {
        lock (Gate)
        {
            try
            {
                if (size == null || capacity < 0) return SQLITE_MISUSE;
                int rc = PathKey(name, out string path);
                if (rc != SQLITE_OK) return rc;
                var node = Find(path);
                if (node == null) return SQLITE_NOTFOUND;
                if (node.References != 0) return SQLITE_BUSY;
                *size = node.Size;
                if (buffer == null) return SQLITE_OK;
                if (capacity < node.Size) return SQLITE_TOOBIG;
                node.Data.AsSpan(0, node.Size).CopyTo(new Span<byte>(buffer, node.Size));
                return SQLITE_OK;
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
    public static int Import(byte* name, void* buffer, long size)
    {
        lock (Gate)
        {
            Node? created = null;
            try
            {
                if (size < 0 || (buffer == null && size != 0)) return SQLITE_MISUSE;
                int rc = PathKey(name, out string path);
                if (rc != SQLITE_OK) return rc;
                var node = Find(path);
                if (node != null && node.References != 0) return SQLITE_BUSY;
                if (node == null) node = created = new Node(path);
                rc = Resize(node, size);
                if (rc != SQLITE_OK) return rc;
                new ReadOnlySpan<byte>(buffer, (int)size).CopyTo(node.Data);
                if (created != null) Nodes.Add(created);
                return SQLITE_OK;
            }
            catch (OutOfMemoryException) { return SQLITE_NOMEM; }
        }
    }
}
